lib/primitives.dart

```diff
@@ -0,0 +1,113 @@
+///     * `pthread_cond_wait` and `pthread_cond_signal`.
+/// * Windows
+///     * [Slim Reader/Writer (SRW) Locks](https://learn.microsoft.com/en-us/windows/win32/sync/slim-reader-writer--srw--locks),
+///     * [Condition Variables](https://learn.microsoft.com/en-us/windows/win32/sync/condition-variables),
+library;
+
+import 'dart:ffi';
+import 'dart:io';
+
+import 'package:ffi/ffi.dart';
+import 'package:native_synchronization/sendable.dart';
+
+import 'package:native_synchronization/src/bindings/pthread.dart';
+import 'package:native_synchronization/src/bindings/winapi.dart';
+
+part 'posix.dart';
+part 'windows.dart';
+
+/// A *mutex* synchronization primitive.
+///
+/// Mutex can be used to synchronize access to a native resource shared between
+/// multiple threads.
+///
+/// [Mutex] object can not be directly sent to other isolates via a `SendPort`,
+/// but it can be converted to a `Sendable<Mutex>` via `asSendable` getter.
+///
+/// Mutex objects are owned by an isolate which created them.
+sealed class Mutex {
+  Mutex._();
+
+  factory Mutex() => Platform.isWindows ? _WindowsMutex() : _PosixMutex();
+
+  /// Acquire exclusive ownership of this mutex.
+  ///
+  /// If this mutex is already acquired then an attempt to acquire it
+  /// blocks the current thread until the mutex is released by the
+  /// current owner.
+  ///
+  /// **Warning**: attempting to hold a mutex across asynchronous suspension
+  /// points will lead to undefined behavior and potentially crashes.
+  void lock();
+
+  /// Release exclusive ownership of this mutex.
+  ///
+  /// It is an error to release ownership of the mutex if it was not
+  /// previously acquired.
+  void unlock();
+
+  /// Run the given [action] while holding exclusive ownership of the
+  /// mutex.
+  R runLocked<R>(R Function() action) {
+    lock();
+    try {
+      return action();
+    } finally {
+      unlock();
+    }
+  }
+
+  Sendable<Mutex> get asSendable => Sendable.wrap(
+      Platform.isWindows ? _WindowsMutex.fromAddress : _PosixMutex.fromAddress,
+      _address);
+
+  int get _address;
+}
+
+/// A *condition variable* synchronization primitive.
+///
+/// Condition variable can be used to synchronously wait for a condition to
+/// occur.
+///
+/// [ConditionVariable] object can not be directly sent to other isolates via a
+/// `SendPort`, but it can be converted to a `Sendable<ConditionVariable>`
+/// object via [asSendable] getter.
+///
+/// [ConditionVariable] objects are owned by an isolate which created them.
+sealed class ConditionVariable {
+  ConditionVariable._();
+
+  factory ConditionVariable() => Platform.isWindows
+      ? _WindowsConditionVariable()
+      : _PosixConditionVariable();
+
+  /// Block and wait until another thread calls [notify].
+  ///
+  /// `mutex` must be a [Mutex] object exclusively held by the current thread.
+  /// It will be released and the thread will block until another thread
+  /// calls [notify].
+  void wait(Mutex mutex);
```

Comment:
Consider also adding timeout support (which is a very common thing to use)

Reply:
Maybe later.